### PR TITLE
docs: add string()

### PR DESCRIPTION
**Pull request by kazizi55**
Implements https://github.com/fabian-hiller/valibot/issues/209#issuecomment-1755624961

### CODE REVIEW

File: website/src/components/Property.tsx
Comment (kazizi55) [inline website/src/components/Property.tsx:1]: badge
copied from modular-form, and rewritten a little bit to avoid using Solid.js
https://github.com/fabian-hiller/modular-forms/blob/69701fec32d0cbf233b42cc7e3080eee060c3bde/website/src/components/Property.tsx#L1-L221

File: website/src/routes/api/(schemas)/string/index.mdx
Comment (kazizi55) [inline website/src/routes/api/(schemas)/string/index.mdx:19]: badge
I think `Pipe` doc is also needed.
Can I write this doc in another PR?

File: website/src/routes/api/(schemas)/string/index.mdx
Comment (kazizi55) [inline website/src/routes/api/(schemas)/string/index.mdx:23]: badge
`StringSchema` doc is also needed, I think.
Can I write this doc in another PR?

File: website/src/routes/api/(schemas)/string/index.mdx
Comment (kazizi55) [inline website/src/routes/api/(schemas)/string/index.mdx:52]: badge
Methods which can be used with string()

File: website/src/routes/api/(schemas)/string/index.mdx
Comment (kazizi55) [inline website/src/routes/api/(schemas)/string/index.mdx:69]: badge
Transformation which can be used with string()

File: website/src/routes/api/(schemas)/string/index.mdx
Comment (kazizi55) [inline website/src/routes/api/(schemas)/string/index.mdx:81]: badge
Validations which can be used with string()

File: website/src/routes/api/(schemas)/string/index.mdx
Comment (fabian-hiller) [reply website/src/routes/api/(schemas)/string/index.mdx:19]: Yes, but wait until I give you feedback on this PR.

File: website/src/routes/api/(schemas)/string/index.mdx
Comment (fabian-hiller) [reply website/src/routes/api/(schemas)/string/index.mdx:23]: The question here is whether we should add this to the `string` API reference as well. Then everything would be in one place. However, it can also make the page messy.

File: website/src/routes/api/(schemas)/string/index.mdx
Comment (fabian-hiller) [reply website/src/routes/api/(schemas)/string/index.mdx:52]: Yes, thats great! Good idea!

File: website/src/routes/api/(schemas)/string/index.mdx
Comment (fabian-hiller) [reply website/src/routes/api/(schemas)/string/index.mdx:69]: Yes, thats great! Good idea!

File: website/src/routes/api/(schemas)/string/index.mdx
Comment (fabian-hiller) [reply website/src/routes/api/(schemas)/string/index.mdx:81]: Yes, thats great! Good idea!

**Comment by fabian-hiller** [comment]
Thank you very much! I'll take a closer look at the PR in the next few days and give you feedback.

File: website/src/routes/api/(schemas)/string/index.mdx
Comment (fabian-hiller) [inline website/src/routes/api/(schemas)/string/index.mdx:13]: Here it is important to me that we represent each overload signature and use the same parameter names as in the code:

> I will probably rename the `error` parameter to `message` in the next days.

```ts
// String schema with pipe
const schema = string(pipe);

// String schema with message and pipe
const schema = string(message, pipe);
```

File: website/src/routes/api/(schemas)/string/index.mdx
Comment (fabian-hiller) [inline website/src/routes/api/(schemas)/string/index.mdx:18]: `<Property />` should support default values via props. Also `message` can also be a function. As with Modular Forms, we should probably include types in the API reference and link them. In this case it would be the [`ErrorMessage`](ot/blob/v0.19.0/library/src/types.ts#L121) type.

File: website/src/routes/api/(schemas)/string/index.mdx
Comment (fabian-hiller) [inline website/src/routes/api/(schemas)/string/index.mdx:19]: The syntax `name: 'Pipe<string>'` is not quite correct:

```ts
const type = {
  type: 'custom',
  name: 'Pipe',
  href: '../Pipe',
  generics: ['string'],
};
```

File: website/src/routes/api/(schemas)/string/index.mdx
Comment (fabian-hiller) [reply website/src/routes/api/(schemas)/string/index.mdx:23]: I have thought about it. Please add the `StringSchema` reference to this PR. You can create a new route for it. I think it is better if everything is in one PR and this PR has the goal to make the `string` API reference perfect.

File: website/src/routes/api/(schemas)/string/index.mdx
Comment (fabian-hiller) [inline website/src/routes/api/(schemas)/string/index.mdx:25]: Here I would make only 2 to 3 concrete examples. We can link the other possibilities below.

```ts
// Schema to validate an eamil
const emailSchema = string([
  minLength(1, 'Please enter your email.'),
  email('The email is badly formatted.'),
  maxLength(30, 'Your email is to long.'),
]);

// Example 2 ...
```

File: website/src/routes/api/(schemas)/string/index.mdx
Comment (fabian-hiller) [reply website/src/routes/api/(schemas)/string/index.mdx:52]: Here I would only add the relevant methods. I would remove `coerceAsync`, `fallbackAsync`, `getDefault`, `parseAsync`, `safeParseAsync` and `unwrap`.

**Review by fabian-hiller — CHANGES_REQUESTED** [review]
Thanks again for your contribution! 🧩

File: website/src/routes/api/(schemas)/string/index.mdx
Comment (fabian-hiller) [reply website/src/routes/api/(schemas)/string/index.mdx:18]: Remember that `message` can also be `undefined`. You can map unions with arrays in `<Property />`:

```tsx
<Property type={['string',  'undefined']} />
```

File: website/src/routes/api/(schemas)/string/index.mdx
Comment (fabian-hiller) [reply website/src/routes/api/(schemas)/string/index.mdx:19]: To not mess up the mdx file, you can also define complex properties at the end of the file like I did in Modular Forms: https://github.com/fabian-hiller/modular-forms/blob/main/website/src/routes/(layout)/[framework]/api/getValues.mdx

Comment (fabian-hiller) [reply website/src/routes/api/(schemas)/string/index.mdx:19]: I would be interested to know how other documentation handles this internally. Maybe we can use that to improve our approach with `<Property />`.

File: website/src/routes/api/(schemas)/string/index.mdx
Comment (fabian-hiller) [reply website/src/routes/api/(schemas)/string/index.mdx:56]: Please also remove `transformAsync`

Comment (kazizi55) [reply website/src/routes/api/(schemas)/string/index.mdx:50]: Fixed!

https://github.com/fabian-hiller/valibot/pull/220/commits/ad5a6d42cc6280673cb9e388eee1baec7eda5793
https://github.com/fabian-hiller/valibot/pull/220/commits/95be95dc84efb7970a2b459717db7a1d16f64d7a

File: website/src/routes/api/(schemas)/string/index.mdx
Comment (kazizi55) [reply website/src/routes/api/(schemas)/string/index.mdx]: Implemented!
https://github.com/fabian-hiller/valibot/pull/220/commits/a224502ecd201223d41674a6d2138702efea9504

File: website/src/routes/api/(schemas)/string/index.mdx
Comment (kazizi55) [reply website/src/routes/api/(schemas)/string/index.mdx]: Implemented!
https://github.com/fabian-hiller/valibot/pull/220/commits/a224502ecd201223d41674a6d2138702efea9504

File: website/src/routes/api/(schemas)/string/index.mdx
Comment (kazizi55) [reply website/src/routes/api/(schemas)/string/index.mdx]: Implemented!
https://github.com/fabian-hiller/valibot/pull/220/commits/a224502ecd201223d41674a6d2138702efea9504

> I would be interested to know how other documentation handles this internally. Maybe we can use that to improve our approach with <Property />.

I'll look this up in the coming days 🙏

**Comment by fabian-hiller** [comment]
You are great! Thanks a lot. I will try to review the changes in the next few days.

One idea I had was to list the authors of the documentation at the bottom of each page, like in Qwik. What do you think about that? Here is an example: https://qwik.builder.io/docs/

**Comment by fabian-hiller** [comment]
@kazizi55 please pull my latest changes, start the website in your development environment and let me know what you think about them.

File: website/src/routes/api/(schemas)/string/index.mdx
Comment (kazizi55) [reply website/src/routes/api/(schemas)/string/index.mdx]: > I would be interested to know how other documentation handles this internally.

I've searched 3 library documents, and found that there are only simple approaches to handle type and property information:
- React Hook Form
  - only code block is used, and there is no component like `<Property />`
  - https://react-hook-form.com/ts#Field
- Material UI
  - only code block and  `` block are used, and there is no component like `<Property />`
  - https://mui.com/material-ui/api/accordion/
- Typia
  - only code block and  `` block are used, and there is no component like `<Property />`
  - https://typia.io/docs/validators/validate/

So as far as I've searched, I guess it might be good to use <Property /> without big changes.

**Comment by kazizi55** [comment]
> Thanks to all the contributors who have helped make this documentation better!

Sounds great!
This would motivate contributors to improve docs more 😄 

Can I add this section in another PR?

File: website/src/routes/api/(schemas)/string/index.mdx
Comment (kazizi55) [inline website/src/routes/api/(schemas)/string/index.mdx:69]: badge
Looks a littile bit messy to me.
Why do you prefer joining links with `,` to with bullet points?
スクリーンショット 2023-11-05 13 26 23

File: website/src/routes/api/(types)/ErrorMessage/index.mdx
Comment (kazizi55) [inline website/src/routes/api/(types)/ErrorMessage/index.mdx:28]: 👍

File: website/src/routes/api/(schemas)/string/index.mdx
Comment (kazizi55) [inline website/src/routes/api/(schemas)/string/index.mdx:32]: 👍

File: website/src/routes/api/(schemas)/string/index.mdx
Comment (fabian-hiller) [reply website/src/routes/api/(schemas)/string/index.mdx]: Thank you for the research!

**Comment by fabian-hiller** [comment]
> Sounds great! This would motivate contributors to improve docs more 😄
> Can I add this section in another PR?

Yes, this would be a big help. Don't focus to much on the design (unless you have experience with it and enjoy doing it). I can improve the CSS afterwards before merging the PR.

File: website/src/routes/api/(schemas)/string/index.mdx
Comment (fabian-hiller) [reply website/src/routes/api/(schemas)/string/index.mdx:69]: I think both are fine, but I haven't finally decided how I want to proceed here. Bullet points seemed a bit long and therefore confusing to me, so I wanted to try this out.

**Comment by fabian-hiller** [comment]
Currently, generics in types cannot be fully mapped with `<Property />`. I noticed this with `StringSchema` for example. We'll have to look at this again.

```
// This is how it looks
BaseSchema<string>

// And this is how it should look
BaseSchema<TOutput = string>
```

File: website/src/routes/api/(schemas)/string/index.mdx
Comment (kazizi55) [reply website/src/routes/api/(schemas)/string/index.mdx:69]: > Bullet points seemed a bit long and therefore confusing to me, so I wanted to try this out.

Got it! 😄

**Comment by kazizi55** [comment]
> Yes, this would be a big help. Don't focus to much on the design (unless you have experience with it and enjoy doing it). I can improve the CSS afterwards before merging the PR.

Thank you for your kindness 😄 
OK! I'm gonna work on it!

**Comment by kazizi55** [comment]
> Currently, generics in types cannot be fully mapped with <Property />. I noticed this with StringSchema for example. We'll have to look at this again.

That's true.
I'm gonna fix it in the next few days 😄

**Comment by kazizi55** [comment]
@fabian-hiller 
> Currently, generics in types cannot be fully mapped with <Property />. I noticed this with StringSchema for example. We'll have to look at this again.

Fixed!
https://github.com/fabian-hiller/valibot/pull/220/commits/2c4c0b9b8303b5ca1b77142391e05b36ec52e353

Now, generics with default parameter can be displayed!
https://deploy-preview-220--valibot.netlify.app/api/StringSchema/

File: website/src/components/Property.tsx
Comment (fabian-hiller) [inline website/src/components/Property.tsx:58]: Can we rename that to `default` so that it is similar to the `default` prop of the component?

Comment (fabian-hiller) [reply website/src/components/Property.tsx:58]: My idea would be to change it to `default?: TypeOrValue`.

Comment (fabian-hiller) [reply website/src/components/Property.tsx:58]: Maybe your approach is the right one. I will probably take a closer look tomorrow.

File: website/src/routes/api/(schemas)/string/index.mdx
Comment (wout-junius) [inline website/src/routes/api/(schemas)/string/index.mdx:34]: Would it be interesting to add 1 example whit the error message?

Comment (fabian-hiller) [reply website/src/routes/api/(schemas)/string/index.mdx:34]: Yes, good idea! Where or how would you add it?

Comment (wout-junius) [reply website/src/routes/api/(schemas)/string/index.mdx:34]: Maybe changing line 52 by
`const UrlSchema = string("A URL must be of type string",[url('The URL is badly formatted.')]);`

**Review by wout-junius — APPROVED** [review]
LGTM

**Comment by fabian-hiller** [comment]
Thanks a lot. I plan to review the new changes in the next days.

**Review by fabian-hiller — APPROVED** [review]

**Comment by fabian-hiller** [comment]
@kazizi55 can you review my changes? I think that we are slowly approaching our destination. 🏁

**Comment by kazizi55** [comment]
@fabian-hiller 
LGTM! Thank you for your changes 😄 
What do you think about the next actions to complete the API document?
Which pages should I write next?

I think it's better to create a new issue which includes to-do list, so that we can check which document is completed or not. 😄

**Comment by fabian-hiller** [comment]
Are you happy with the current version? Would you change anything? I will take another look at everything in the next few days and also check how other API references present there information to give feedback. But we are on the right track and I think we will reach a final version soon.

**Comment by fabian-hiller** [comment]
I have further improved the `string` API reference and updated the references of all types. I'll take another look at it tomorrow. I think we can merge the changes shortly and then add the other API references one by one. 🚀

**Comment by fabian-hiller** [comment]
@kazizi55 thank you again for your contribution! I will merge this PR and post the draft on Twitter to get more feedback before we add the other references.

**Comment by fabian-hiller** [comment]
https://valibot.dev/api/string/

**Comment by fabian-hiller** [comment]
Unfortunately, no. We probably only get feedback when people use the API reference and don't understand or miss something. I will create a issue soon where we can coordinate the API reference extension. 🧩

**Pull request merged**
